ecosystem/indexer-grpc/indexer-grpc-table-info/src/backup_restore/gcs.rs

```diff
@@ -0,0 +1,293 @@
+                    object: METADATA_FILE_NAME.to_string(),
+                    ..Default::default()
+                },
+                &Range::default(),
+            )
+            .await
+            .map(BackupRestoreMetadata::from)
+    }
+
+    pub async fn update_metadata(&self, chain_id: u64, epoch: u64) -> anyhow::Result<()> {
+        let metadata = BackupRestoreMetadata::new(chain_id, epoch);
+        loop {
+            match self
+                .gcs_client
+                .upload_object(
+                    &UploadObjectRequest {
+                        bucket: self.bucket_name.clone(),
+                        ..Default::default()
+                    },
+                    serde_json::to_vec(&metadata).unwrap(),
+                    &UploadType::Simple(Media {
+                        name: Borrowed(METADATA_FILE_NAME),
+                        content_type: Borrowed(JSON_FILE_TYPE),
+                        content_length: None,
+                    }),
+                )
+                .await
+            {
+                Ok(_) => {
+                    return Ok(());
+                },
+                // https://cloud.google.com/storage/quotas
+                // add retry logic due to: "Maximum rate of writes to the same object name: One write per second"
+                Err(Error::Response(err)) if (err.is_retriable() && err.code == 429) => {
+                    info!("Retried with rateLimitExceeded on gcs single object at epoch {} when updating the metadata", epoch);
+                    tokio::time::sleep(Duration::from_millis(500)).await;
+                    continue;
+                },
+                Err(err) => {
+                    anyhow::bail!("Failed to update metadata: {}", err);
+                },
+            }
+        }
+    }
+
+    pub async fn backup_db_snapshot(
+        &self,
+        chain_id: u64,
+        epoch: u64,
+        indexer_async_v2: Arc<IndexerAsyncV2>,
+        snapshot_path: PathBuf,
+    ) -> anyhow::Result<()> {
+        // reading epoch from gcs metadata is too slow, so updating the local var first so that every previous
+        // new epoch based backup will be observed correctly by the next backup
+        self.set_metadata_epoch(epoch);
+
+        // rocksdb will create a checkpoint to take a snapshot of full db and then save it to snapshot_path
+        indexer_async_v2
+            .create_checkpoint(&snapshot_path)
+            .context(format!("DB checkpoint failed at epoch {}", epoch))?;
+
+        // Spawn blocking a thread to create a gzipped tar file by compressing a folder into a single file
+        // synchronously without blocking the async thread
+        let snapshot_path_closure = snapshot_path.clone();
+        let (tar_file, _tar_file_name) = task::spawn_blocking(move || {
+            create_tar_gz(snapshot_path_closure.clone(), &epoch.to_string())
+                .expect("Failed to create tar.gz file in blocking task")
```

Comment:
why panicking within spawn_blocking, instead of "this thread"? (then you don't need the closure?)